### ISSUE TITLE
property in generic conditional type not narrowed in nested conditional type

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#46145]
# Bug Report

### 🕗 Version & Regression Information
4.5.0-dev.20210930

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.5.0-dev.20210930#code/JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyyMmAXMgM5hSgDmAPgK4gAmEMoELBAvgQWACeABxQBlagB4AKsggAPSKwqVqdAHzIAvMmn98Q0cgBCcKDM06CRWQqUsVGdNaLIA-LoDaAchLpvALpyihDKqjQgtC6uyAD0sboiKN7SngBEfmkB3sgs6BAqIOhglHBgwBQwgshgABYoCOggVFBwoCXeLXTeAHTRrvE2ScidapHI9MjMbBwgXDkVyEUlcBQUwLQgcABGADYoYFiGyV2Rvf1Eg0NG3tPsnCwLhcXIq+ubO-s1R8OjEbS9AAUACYAMwAFnBAEoLu5kBJzKlfJhAupYeQ5gA3aD9UhY6BAA)

### 💻 Code

```ts
interface Foo {
  foo: string|undefined
}

type Str<T extends string> = T

type Bar<T> = 
  T extends Foo
    ? T['foo'] extends string
      // Type 'T["foo"]' does not satisfy the constraint 'string'.
      //  Type 'string | undefined' is not assignable to type 'string'.
      //   Type 'undefined' is not assignable to type 'string'.(2344)
      ? Str<T['foo']>
      : never
    :never
```

### 🙁 Actual behavior
`T['foo']` is not being narrowed to `string`

### 🙂 Expected behavior
`T['foo']` is narrowed to `string`

**Comment by craigphicks** [comment]
@DetachHead 
I agree it is not a correct compiler response.  

However, if there are other paths that lead to the target, will that suffice?

E.g., if the redundant `Str<...>` is omitted it seems to work: 

```
interface Foo {
  foo: string|undefined
}
// type Str<T extends string> = T
type Bar<T> = 
  T extends Foo
    ? T['foo'] extends string 
      // Type 'T["foo"]' does not satisfy the constraint 'string'.
      //  Type 'string | undefined' is not assignable to type 'string'.
      //   Type 'undefined' is not assignable to type 'string'.(2344)
      ? T['foo']
      : "never1"
    :"never2"

const bar1:Bar<{foo:string}>="1";
const bar2:Bar<{foo:"1"}>="1";

type BarX1 = Bar<{}>; // type BarX1 = "never1"
type BarX2 = Bar<{foo:undefined}>; // type BarX2 = "never2"
```

Also, the output of Bar can be passed to Str as follows:
```
type Str<T extends string> = T

type Bar1<T> = 
  T extends Foo
    ? T['foo'] extends string 
      // Type 'T["foo"]' does not satisfy the constraint 'string'.
      //  Type 'string | undefined' is not assignable to type 'string'.
      //   Type 'undefined' is not assignable to type 'string'.(2344)
      ? T['foo']
      : never
    :never


const str1:Str<Bar<{foo:string}>>="1";
const str2:Str<Bar<{foo:"1"}>>="1";

type StrX1 = Str<Bar1<{}>>; // type StrX1 = never
type StrX2 = Str<Bar1<{foo:undefined}> // type StrX2 = never
```

[playground](https://www.typescriptlang.org/play?ts=4.5.0-dev.20210930#code/JYOwLgpgTgZghgYwgAgGIHt3IN4FgBQyyMmAXMgM5hSgDmAPgK4gAmEMoELBAvgQPT9kYAJ4AHFAGVqAHgAqyCAA9IrCpWp0AfMgC8yOQVETkAIThR5O-QSILlqluozpbRZAH4DAbQDkJdF8AXUUVCDUNGhBaZDd3ZEEDcRRfOW8AIgD0oN9kFnQIdRB0MEo4MGAKGBFhAAsUBHQQKig4UFLfFrpfADo490Skk07NaOR6ZGY2DhAuXMrkYtK4CgpgWhA4ACMAGxQwLGMUruje-qJBoZSp9k4WeaKS5BW1je294UPk5BGo2l6ABQAJgAzAAWMEASnOnh8-kwwRh5HSswAbtAAIzpfqkFEQdFQIHY-AERrNUpbCwY0jmSzYAKkE60HhaXTpLEAblJTSoyEphJpFhk9LI7PSLLZnIIRm+tIAGhi9GYhdgWRyEkIjsqoAqlXiCViZSZ5UClbThQybjMuGqNcJZRY5ab9ProESgA)

Maybe your point is worth solving even if there are other paths to the target.
Or Maybe you simplified a bigger problem for posting and so the above isn't applicable to that bigger problem.

So please just consider this post as lateral thoughts.

**Comment by DetachHead** [comment]
> Or Maybe you simplified a bigger problem for posting and so the above isn't applicable to that bigger problem.

Yeah, here it is if you're curious https://github.com/DetachHead/typetime-json-schema/blob/master/src/index.ts#L45

**Comment by craigphicks** [comment]
As far a workaround goes you might try to manually extract the type:
```
interface Foo  {
  foo: string|undefined
};
type Str<T extends string> = T

// manually removing `undefined` from T['foo'] works
type Bar1<T> = 
  T extends Foo
    ? T['foo'] extends string 
      ?Str<Extract<T['foo'],string>>    // Extract
      : never
    :never

const str1:Str<Bar1<{foo:string}>>="1";
const str2:Str<Bar1<{foo:"1"}>>="1";
type BarX1 = Bar1<{}>; // type BarX1 = never
type BarX2 = Bar1<{foo:undefined}>; // type BarX2 = never
```
[playground](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.4.3#code/JYOwLgpgTgZghgYwgAgGIHt3OQbwLABQ2MmAXMgM5hSgDmAPgK4gAmEMoELhAvgNyEwATwAOKAMrUAPABVkEAB6RWFStToA+ZAF5kMwoQD0h5AFs4IRnAA21ocigRT6AG51kAA2ZsOILh+QYKHRTPQBtAHISdAiAXWQAd3QoAGsKQVEUACE4KABGWS1dQmw5RWUWVQx0EuxkAH5wqMw4+SUIFTUaEFpkWrqGySgpAFElKEQwWUjouIAaKm7aDQ1+uvI-F2g10k3tgkIEdBAqLrzSIakc-KkcaNJFuh4V7QAiPNeBAiOTsC6AJgu0muBTuZHer2eGjeHy+wjEyGuAA08jpEblQc8+MhjMh4dlcii0XsoBkEcj-miQbd7t52JwWFicSZ8eioEjKboSQYCEA)

But again, that's just a lateral comment.  It doesn't deprecate the severity of your point.

**Comment by craigphicks** [comment]
Out of curiosity, I tried setting 'exactOptionalPropertyTypes' to true in TS Config and redefining Foo's foo property to be optional:

```
// exactOptionalPropertyTypes is set to true in TS Config
type Foo = {
  foo?: string
};
type Str<T extends string> = T

type Bar<T> = 
  T extends Foo
    ? T['foo'] extends string 
      // Type 'T["foo"]' does not satisfy the constraint 'string'.
      //  Type 'string | undefined' is not assignable to type 'string'.
      //   Type 'undefined' is not assignable to type 'string'.(2344)
      ?Str<T['foo']>
      : never
    :never
```
In this case it seems that  'undefined` got forcibly added at some point, because it wasn't there to begin with.

**Comment by andrewbranch** [comment]
I think this is a duplicate of https://github.com/microsoft/TypeScript/issues/45651. I’m going to mark it that way and add this as an additional test when addressing that one.
